### PR TITLE
ENH: Refine connection exception handling

### LINKED ISSUE #110: Inappopriate assert
I suspect there's scanner to send malformed data to the expose port from internet,  may remove this assert and change to warning

```
2024-11-06 11:01:46,326 asyncio      7 ERROR    Task exception was never retrieved
future: <Task finished name='Task-102' coro=<SocketServer.create.<locals>.handle_connection() done, defined at /usr/local/lib/python3.10/dist-packages/xoscar/backends/communication/socket.py:239> exception=Asser
tionError('Received malformed data, please check Xoscar version on both side.')>
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/xoscar/backends/communication/socket.py", line 241, in handle_connection
    return await server.on_connected(
  File "/usr/local/lib/python3.10/dist-packages/xoscar/backends/communication/socket.py", line 159, in on_connected
    await self.channel_handler(channel)
  File "/usr/local/lib/python3.10/dist-packages/xoscar/backends/pool.py", line 557, in on_new_channel
    message = await channel.recv()
  File "/usr/local/lib/python3.10/dist-packages/xoscar/backends/communication/socket.py", line 90, in recv
    header = await deserializer.get_header()
  File "/usr/local/lib/python3.10/dist-packages/xoscar/serialization/aio.py", line 138, in get_header
    return cloudpickle.loads(await self._get_obj_header_bytes())
  File "/usr/local/lib/python3.10/dist-packages/xoscar/serialization/aio.py", line 109, in _get_obj_header_bytes
    header_length = get_header_length(header_bytes)
  File "/usr/local/lib/python3.10/dist-packages/xoscar/serialization/aio.py", line 80, in get_header_length
    assert version == DEFAULT_SERIALIZATION_VERSION, MALFORMED_MSG
AssertionError: Received malformed data, please check Xoscar version on both side.

```

### PR DESCRIPTION
**Pull request by frostyplanet**
## What do these changes do?

While server-side recv invalid format data, previously will not handle exception and leaving connection not close.

The exception situation including invalid header,  or deserialization fails

## Related issue number

Fixes #110

## Check code requirements

None

### CODE REVIEW

File: python/xoscar/backends/pool.py [hunk 1]
```diff
@@ -554,7 +554,8 @@ async def on_new_channel(self, channel: Channel):
         while not self._stopped.is_set():
             try:
                 message = await channel.recv()
-            except EOFError:
+            except Exception as e:
```
Comment (qinxuye) [inline python/xoscar/backends/pool.py:557]: Better to specify which errors clearly.

File: python/xoscar/backends/core.py [hunk 2]
```diff
@@ -74,15 +74,22 @@ async def _listen(self, client: Client):
...
                 try:
                     message: _MessageBase = await client.recv()
-                except (EOFError, ConnectionError, BrokenPipeError):
+                except (
+                    EOFError,
+                    ConnectionError,
+                    BrokenPipeError,
+                    AssertionError,
+                ) as e:
+                    # AssertionError is from get_header
```
Comment (qinxuye) [inline python/xoscar/backends/core.py:83]: Maybe we should raise ConnectionError or whatever for `get_header`.

File: python/xoscar/backends/pool.py (hunk 1 above)
Comment (frostyplanet) [reply python/xoscar/backends/pool.py:557]: Should close the connection on all errors,  otherwise socket will not be gc, making a fd leak

**Review by qinxuye — APPROVED** [review]
LGTM

**Pull request merged**